alpenhorn/io/default.py

```diff
@@ -163,6 +218,109 @@ def md5(self, path: str | pathlib.Path, *segments) -> str:
         """
         return util.md5sum_file(pathlib.Path(self.node.root, path, *segments))
 
+    def pull(self, req: ArchiveFileCopyRequest) -> None:
+        """Pull file specified by copy request `req` onto `self.node`.
+
+        Most of the work happens in an asynchronous I/O task.
+
+        Parameters
+        ----------
+        req : ArchiveFileCopyRequest
+            the copy request to fulfill.  We are the destination node (i.e.
+            `req.group_to == self.node.group`).
+        """
+
+        if self.node.under_min:
+            log.info(
+                f"Skipping pull for StorageNode f{self.node.name}: hit minimum free space: "
+                f"({self.node.avail_gb:.2f} GiB < {self.node.min_avail_gb}:.2f GiB)"
+            )
+            return
+
+        if self.node.check_over_max():
+            log.info(
+                f"Skipping pull for StorageNode f{self.node.name}: node full. "
+                f"({self.node.get_total_gb():.2f} GiB >= {self.node.max_total_gb}:.2f GiB)"
+            )
+            return
+
+        # Check that there is enough space available (and reserve what we need)
+        if not self.reserve_bytes(req.file.size_b):
```

Comment:
I might not be understanding correctly but aren't these space checks somewhat redundant? I would think that just checking `self.reserve_bytes` would be enough. At the very least, if `under_min` passes then `over_max` would also pass, so I'm not sure if both are needed. Or are the checks just done for proper logging of the issue?

Reply:
The tests are inter-related, but I don't think they're actually redundant.

The `over_max` and `under_min` tests are _database-only_ checks, which are not at all affected by the size of the file being transfered. They return the results of:
 * over max: `StorageNode.max_total_gb >= StorageNode.avail_gb`
 * under min: `StorageNode.min_avail_gb <= StorageNode.avail_gb`

The value of `avail_gb` _does_ ultimately get updated by checking the filesystem, but not during the check itself. 

The values of `max_total_gb` and `min_avail_gb` are set manually (when desired) to adjust how alpenhorn works:
* `max_total_gb` prevents a StorageNode from taking up more space than allotted to it.  We use it, for instance, on the `cedar_online` node to prevent the raw data from using all of our `/project` space.
* `min_avail_gb` prevents a StorageNode from using up all available space on a disk.  We could use it, for instance to ensure some minimum amount of space remains free on `/project`.

(On a filesystem which is used _only_ as a storage node, there _is_ a degeneracy here: setting `max_total_gb` to a value, _V_ is equivalent to setting `min_avail_gb` to: _(size-of-filesystem) - V_.)

Neither of these parameters need to be set (and, for most nodes they aren't), in which case these two checks automatically succeed.

Conversely, while `over_min` and `under_max` are database checks, the `reserve_bytes` check is an check against the filesystem (an I/O check).  It is meant to, regardless of database constraints we've put on a storage node, determine whether a given file is going to fit on the underlying filesystem.  (In cases where a single file is larger than the value of `min_avail_gb`, the `under_min` check can succeed while reservation can fail.)

The reserve bytes check is also necessary to prevent multiple pull jobs overfilling the filesystem by each job thinking there's enough space by not considering all the other pull jobs concurrently happening.

Reply:
That clears it up, thanks!